### PR TITLE
Virtually split the repeats of repeated shards

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

Repeat the samples of a shard as [ABC][ABC][AB] instead of [AAABBBCC], i.e. break repeated shards into (number of full and partial repeats) ephemeral shards for the purposes of shuffling.

Most shuffle algorithms have a first stage that globally shuffles the list of shards. This change will cause each repeat of an upsampled shard to become scattered differently throughout the epoch, leading to more even shuffling, better than resorting to very large shuffle block sizes.

Secondly, we include a script to visualize shuffle behavior. Example below:

shuffle_plot

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: scripts/shuffle/vis.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--num_shards', type=int, default=100)
+    args.add_argument('--base_samples_per_shard', type=int, default=20)
+    args.add_argument('--num_canonical_nodes', type=int, default=16)
+    args.add_argument('--seed', type=int, default=0xD06E)
```
Comment (karan6181) [inline scripts/shuffle/vis.py:24]: seed default value !!! is it Hexa-demical value ?

File: tests/test_mixing.py [hunk 2]
```diff
@@ -180,8 +181,24 @@ def test_mix_proportion_range(root: str):
...
     assert dataset.num_samples == 8
-    assert walk(dataset) == [2, 3, 4, 4, 5, 5, 6, 6, 6, 7, 7, 7]
+    assert walk(dataset) == [2, 3, 4, 5, 4, 5, 6, 7, 6, 7, 6, 7]
     for i, stream in enumerate(streams):
         assert float_eq(stream.proportion, i / 6)
         assert stream.repeat == i
         assert stream.samples == i * 2
+
+
+def test_mix_balance(root: str):
```
Comment (karan6181) [inline tests/test_mixing.py:191]: You can use `local_remote_dir` pytest annotation which deletes the directory when test finishes. checkout `test_writer.py` for example

**Review by karan6181 — COMMENTED** [review]
Looks good overall. Apart from unit test, have you also tested the sample ordering on a larger number of sample ?

File: scripts/shuffle/vis.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 MosaicML Streaming authors
```
Comment (dakinggg) [inline scripts/shuffle/vis.py:1]: Can you include the visual comparison in the pr description?

File: tests/test_mixing.py (hunk 2 above)
Comment (knighton) [reply tests/test_mixing.py:191]: done

File: scripts/shuffle/vis.py (hunk 1 above)
Comment (knighton) [reply scripts/shuffle/vis.py:24]: yep

**Comment by dakinggg** [comment]
Also, should we get Jonathan to do a 1b run under the same settings that he saw the shuffle issues before and make sure this fixes thing at least as well as the py1b shuffle fix?

**Comment by knighton** [comment]
shuffle_plot

File: scripts/shuffle/vis.py (hunk 3 above)
Comment (knighton) [reply scripts/shuffle/vis.py:1]: done

**Review by karan6181 — APPROVED** [review]
LGTM!

**Pull request merged**